src/Dictionaries/CacheDictionary.h

```diff
@@ -198,49 +198,41 @@ class CacheDictionary final : public IDictionary
     template <typename Value>
     using ContainerPtrType = std::unique_ptr<ContainerType<Value>>;
 
+    using time_point_t = std::chrono::system_clock::time_point;
+
     struct CellMetadata final
     {
-        using time_point_t = std::chrono::system_clock::time_point;
-        using time_point_rep_t = time_point_t::rep;
-        using time_point_urep_t = std::make_unsigned_t<time_point_rep_t>;
-
-        static constexpr UInt64 EXPIRES_AT_MASK = std::numeric_limits<time_point_rep_t>::max();
-        static constexpr UInt64 IS_DEFAULT_MASK = ~EXPIRES_AT_MASK;
-
         UInt64 id;
-        /// Stores both expiration time and `is_default` flag in the most significant bit
-        time_point_urep_t data;
+        time_point_t deadline;
+        bool is_default{false};
 
-        time_point_t strict_max;
+        time_point_t expiresAt() const { return deadline; }
+        void setExpiresAt(const time_point_t & t) { deadline = t; is_default = false; }
+        bool isDefault() const { return is_default; }
+        void setDefault() { is_default = true; }
+    };
 
-        /// Sets expiration time, resets `is_default` flag to false
-        time_point_t expiresAt() const { return ext::safe_bit_cast<time_point_t>(data & EXPIRES_AT_MASK); }
-        void setExpiresAt(const time_point_t & t) { data = ext::safe_bit_cast<time_point_urep_t>(t); }
+    using AttributeValue = std::variant<
+        UInt8, UInt16, UInt32, UInt64, UInt128,
+        Int8, Int16, Int32, Int64,
+        Decimal32, Decimal64, Decimal128,
+        Float32, Float64, String>;
 
-        bool isDefault() const { return (data & IS_DEFAULT_MASK) == IS_DEFAULT_MASK; }
-        void setDefault() { data |= IS_DEFAULT_MASK; }
+    struct AttributeValuesForKey
+    {
+        bool found{false};
+        std::vector<AttributeValue> values;
+
+        std::string dump();
     };
 
+    using FoundValuesForKeys = std::unordered_map<Key, AttributeValuesForKey>;
+
     struct Attribute final
     {
         AttributeUnderlyingType type;
-        std::variant<
-            UInt8,
-            UInt16,
-            UInt32,
-            UInt64,
-            UInt128,
-            Int8,
-            Int16,
-            Int32,
-            Int64,
-            Decimal32,
-            Decimal64,
-            Decimal128,
-            Float32,
-            Float64,
-            String>
-            null_values;
+        String name;
+        AttributeValue null_values;
```

Comment:
Btw, what are null_values and especially arrays? null_values is probably a pre-computed null value of a given field type? Unsure what to think about arrays. The plural suffixes are somewhat misleading.
A comment would be helpful here.

Reply:
I've wrote a comment about it. null_value is a default value for an attribute, can be defined in config. "arrays" are values of this attribute for all keys. It is very fast, I tried to implement something else, but it was slower.

Reply:
Should we call `null_value` `default_value` then?

Reply:
Of course we can, but we have "null value" name in [config file](https://clickhouse.tech/docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-structure/#ext_dict_structure-attributes). I think it will be a little discrepancy for new readers.